File: xenon-android-library/src/main/kotlin/com/abubusoft/xenon/mesh/tiledmaps/ObjectLayer.java
package com.abubusoft.xenon.mesh.tiledmaps;

import java.util.ArrayList;

import com.abubusoft.xenon.mesh.tiledmaps.internal.AbstractLayerHandler;
import com.abubusoft.xenon.mesh.tiledmaps.internal.TiledMapView;
import com.abubusoft.xenon.mesh.tiledmaps.internal.ObjectLayerHandler;
import com.abubusoft.xenon.mesh.tiledmaps.internal.LayerDrawer;
import com.abubusoft.xenon.mesh.tiledmaps.tmx.loader.SAXUtil;
import com.abubusoft.xenon.mesh.tiledmaps.tmx.loader.TMXPredefinedProperties;
import org.xml.sax.Attributes;

public class ObjectLayer extends Layer {

	/**
	 * <p>
	 * Elenco di oggetti presenti nel layer.
	 * </p>
	 */
	public ArrayList<ObjDefinition> objects;

	/**
	 * <p>
	 * Drawer da utilizzare per il draw del layer.
	 * </p>
	 */
	public ObjectLayerDrawer objectDrawer;

	private ObjectLayerHandler handler;

	/**
	 * <p>
	 * Costruttore
	 * </p>
	 * 
	 * @param tiledMap
	 * @param atts
	 */
	public ObjectLayer(TiledMap tiledMap, Attributes atts) {
		super(LayerType.OBJECTS, tiledMap, atts);

		objects = new ArrayList<ObjDefinition>();
		visible = SAXUtil.getInt(atts, TMXPredefinedProperties.VISIBLE, 1) == 1;
	}

	/**
	 * <p>
	 * Elenco di oggetti definiti nel layer
	 * </p>
	 * 
	 * @return elenco degli oggetti
	 */
	public ArrayList<ObjDefinition> getObjects() {
		return objects;
	}

	/**
	 * <p>
	 * Aggiunge un oggetto.
	 * </p>
	 * 
	 * @param object
	 */
	public void addObject(ObjDefinition object) {
		objects.add(object);
	}

	/* (non-Javadoc)
	 * @see com.abubusoft.xenon.mesh.tiledmaps.Layer#onBuildView(com.abubusoft.xenon.mesh.tiledmaps.internal.MapView)
	 */
	@Override
	public void onBuildView(TiledMapView view) {
		handler.onBuildView(view);
	}

	/**
	 * <p>
	 * Imposta il drawer da utilizzare per il rendering.
	 * </p>
	 * 
	 * @param value
	 */
	public void setObjectDrawer(ObjectLayerDrawer value) {
		objectDrawer = value;
	}
	
	@Override
	protected void buildHandler(AbstractLayerHandler<?> handler) {
		this.handler=(ObjectLayerHandler) handler;		
	}
	
	@Override
	public LayerDrawer drawer() {
		return handler;
	}

	@Override
	public TiledMapView view() {
		return handler.view();
	}
}

File: xenon-android-library/src/main/kotlin/com/abubusoft/xenon/mesh/tiledmaps/ObjectLayer.kt
package com.abubusoft.xenon.mesh.tiledmaps

import com.abubusoft.xenon.mesh.tiledmaps.internal.AbstractLayerHandler
import com.abubusoft.xenon.mesh.tiledmaps.internal.LayerDrawer
import com.abubusoft.xenon.mesh.tiledmaps.internal.ObjectLayerHandler
import com.abubusoft.xenon.mesh.tiledmaps.internal.TiledMapView
import com.abubusoft.xenon.mesh.tiledmaps.tmx.loader.SAXUtil.getInt
import com.abubusoft.xenon.mesh.tiledmaps.tmx.loader.TMXPredefinedProperties
import org.xml.sax.Attributes

class ObjectLayer(tiledMap: TiledMap, atts: Attributes?) : Layer(LayerType.OBJECTS, tiledMap, atts) {
    /**
     *
     *
     * Elenco di oggetti definiti nel layer
     *
     *
     * @return elenco degli oggetti
     */
    /**
     *
     *
     * Elenco di oggetti presenti nel layer.
     *
     */
    var objects: ArrayList<ObjDefinition>

    /**
     *
     *
     * Drawer da utilizzare per il draw del layer.
     *
     */
    var objectDrawer: ObjectLayerDrawer? = null
    private var handler: ObjectLayerHandler? = null

    /**
     *
     *
     * Costruttore
     *
     *
     * @param tiledMap
     * @param atts
     */
    init {
        objects = ArrayList()
        visible = getInt(atts!!, TMXPredefinedProperties.VISIBLE, 1) == 1
    }

    /**
     *
     *
     * Aggiunge un oggetto.
     *
     *
     * @param object
     */
    fun addObject(`object`: ObjDefinition) {
        objects.add(`object`)
    }

    /* (non-Javadoc)
	 * @see com.abubusoft.xenon.mesh.tiledmaps.Layer#onBuildView(com.abubusoft.xenon.mesh.tiledmaps.internal.MapView)
	 */
    override fun onBuildView(view: TiledMapView) {
        handler!!.onBuildView(view)
    }

    /**
     *
     *
     * Imposta il drawer da utilizzare per il rendering.
     *
     *
     * @param value
     */
    fun setObjectDrawer(value: ObjectLayerDrawer?) {
        objectDrawer = value
    }

    override fun buildHandler(handler: AbstractLayerHandler<*>?) {
        this.handler = handler as ObjectLayerHandler?
    }

    override fun drawer(): LayerDrawer? {
        return handler
    }

    override fun view(): TiledMapView? {
        return handler!!.view()
    }
}